### PR TITLE
HIVE-24203: Implement stats annotation rule for the LateralViewJoinOperator

### PR DESCRIPTION
**Pull request by okumin**
### What changes were proposed in this pull request?

Estimate statistics of LATERAL VIEW correctly.

StatsRulesProcFactory doesn't have any rules to handle a JOIN by LATERAL VIEW.
This can cause an underestimation in case that UDTF in LATERAL VIEW generates multiple rows.

### Why are the changes needed?

Significant underestimation can happen when LATERAL VIEW increases the number of records a lot and the source table has large.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added one test case.

### CODE REVIEW

File: ql/src/test/queries/clientpositive/annotate_stats_lateral_view_join.q [hunk 1]
```diff
@@ -0,0 +1,38 @@
+set hive.fetch.task.conversion=none;
```
Comment (okumin) [inline ql/src/test/queries/clientpositive/annotate_stats_lateral_view_join.q:1]: To make EXPLAIN show Statistics. I'm thinking to create another ticket and add this line to other `annotate_stats_*.q`.

e.g. https://github.com/apache/hive/blob/master/ql/src/test/results/clientpositive/llap/annotate_stats_select.q.out

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java [hunk 2]
```diff
@@ -2921,6 +2920,77 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
     }
   }
 
+  /**
+   * LateralViewJoinOperator joins the output of select with the output of UDTF.
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:2924]: Could you provide a bit more details about what the rule does? Most of the other rules in this class give a general overview of the cost model they implement.

Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java]: @zabetak Thanks for taking a look!
I added a description. Please feel free to ask me if something doesn't make sense.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java [hunk 3]
```diff
@@ -2921,6 +2920,97 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
...
+      }
+
+      final List<Operator<? extends OperatorDesc>> parents = lop.getParentOperators();
+      if (parents.size() != 2) {
+        LOG.warn("LateralViewJoinOperator should have just two parents but actually has "
+                + parents.size() + " parents.");
+        return null;
+      }
+
+      final Statistics selectStats = parents.get(LateralViewJoinOperator.SELECT_TAG).getStatistics().clone();
```
[the remark this answers is not in the document]
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:2961]: As for `udtfStats`, we can totally avoid clone.
As for `udtfStats`, its column stats will be updated. However, looks like `StatsUtils.getColStatisticsFromExprMap` clones them?
Anyway I think we can remove them if CI passes. I will try it.

File: ql/src/test/results/clientpositive/llap/annotate_stats_lateral_view_join.q.out [hunk 4]
```diff
@@ -503,14 +503,14 @@ STAGE PLANS:
                             Statistics: Num rows: 1 Data size: 376 Basic stats: COMPLETE Column stats: COMPLETE
                             Lateral View Join Operator
                               outputColumnNames: _col0, _col1, _col5, _col6
-                              Statistics: Num rows: 0 Data size: 24 Basic stats: PARTIAL Column stats: NONE
+                              Statistics: Num rows: 0 Data size: 24 Basic stats: PARTIAL Column stats: COMPLETE
```
Comment (okumin) [inline ql/src/test/results/clientpositive/llap/annotate_stats_lateral_view_join.q.out:758]: With clone, the following condition is not satisfied since the basic stats of parent operators are PARTIAL.
https://github.com/apache/hive/blob/91e492de239427fc1e38e5e4350cfdce409ebb70/ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java#L2969

Comment (okumin) [reply ql/src/test/results/clientpositive/llap/annotate_stats_lateral_view_join.q.out:758]: BTW, it would be better that the UDTF rule puts one in num rows in case that it becomes zero.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java]: Nothing was not unexpectedly broken. CI failed but it would not be related to this PR...
- https://github.com/apache/hive/pull/1531/commits/91e492de239427fc1e38e5e4350cfdce409ebb70

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java [hunk 5]
```diff
@@ -2921,6 +2920,97 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
...
+    }
+
+    private List<ColStatistics> multiplyColStats(List<ColStatistics> colStatistics, double factor) {
+      for (ColStatistics colStats : colStatistics) {
+        colStats.setNumFalses(StatsUtils.safeMult(colStats.getNumFalses(), factor));
+        colStats.setNumTrues(StatsUtils.safeMult(colStats.getNumTrues(), factor));
+        colStats.setNumNulls(StatsUtils.safeMult(colStats.getNumNulls(), factor));
+        // When factor > 1, the same records are duplicated and countDistinct never changes.
+        if (factor < 1.0) {
+          colStats.setCountDistint(StatsUtils.safeMult(colStats.getCountDistint(), factor));
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:3007]: I think we should make sure that NDV is at least 1 in case numrows is >0

File: ql/src/test/results/clientpositive/llap/annotate_stats_lateral_view_join.q.out (hunk 4 above)
Comment (kgyrtkirk) [reply ql/src/test/results/clientpositive/llap/annotate_stats_lateral_view_join.q.out:758]: definetly - I don't think it will be `0` in reality!

Comment (okumin) [reply ql/src/test/results/clientpositive/llap/annotate_stats_lateral_view_join.q.out:758]: This is an edge case since `HIVE_STATS_UDTF_FACTOR` is greater than or equal to 1. Anyway, I created a ticket.
https://issues.apache.org/jira/browse/HIVE-24240

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java (hunk 5 above)
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:3007]: Now I think this is available for this purpose if we add updating num trues and num falses?
https://github.com/apache/hive/blob/c082a724648b6bfbdd4b0ff72d7c41c29257beba/ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java#L2050-L2100

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java [hunk 6]
```diff
@@ -2921,6 +2920,97 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
...
+        final List<ColStatistics> udtfColStats = StatsUtils
+                .getColStatisticsFromExprMap(conf, udtfStats, columnExprMap, schema);
+        joinedStats.addToColumnStats(udtfColStats);
+
+        joinedStats = applyRuntimeStats(aspCtx.getParseContext().getContext(), joinedStats, lop);
+        lop.setStatistics(joinedStats);
+
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("[0] STATS-" + lop.toString() + ": " + joinedStats.extendedToString());
+        }
```
[the remark this answers is not in the document]
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:2988]: I wonder if we should switch `[0]` or `[1]` based on a condition. I can see some rules use a different marker based on maybe the existence of column stats.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java (hunk 5 above)
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:3007]: This method may include additional logging and logics to optimize JOIN such as `cs.setFilterColumn`. It would be better to implement a simple and separate utility.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:2988]: I don't know what's the point of these `[0]`/`[1]` markers; from one of the historical commits it seems to me like these are some kind of "log message indexes" inside the method ....
I think we could stop doing that...

[the remark this answers is not in the document]
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java]: I also agree and I did that.
https://github.com/apache/hive/pull/1531/commits/d333d5d70184a1cf1f0c0f239e9229965e486202

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java [hunk 7]
```diff
@@ -2921,6 +2920,97 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
...
+      if (parents.size() != 2) {
+        LOG.warn("LateralViewJoinOperator should have just two parents but actually has "
+                + parents.size() + " parents.");
+        return null;
+      }
+
+      final Statistics selectStats = parents.get(LateralViewJoinOperator.SELECT_TAG).getStatistics();
+      final Statistics udtfStats = parents.get(LateralViewJoinOperator.UDTF_TAG).getStatistics();
+
+      final double factor = (double) udtfStats.getNumRows() / (double) selectStats.getNumRows();
```
[the remark this answers is not in the document]
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java]: Added steps to check both numbers and ensure at least one record on stats.
https://github.com/apache/hive/pull/1531/commits/50396346eaed5d6bab4ff87dd079918a769a7ebd

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java (hunk 5 above)
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java]: Ceiled. I moved this method since I'd like to reuse it for HIVE-24240.
https://github.com/apache/hive/commit/50396346eaed5d6bab4ff87dd079918a769a7ebd

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java [hunk 8]
```diff
@@ -2961,10 +2961,11 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
       final Statistics selectStats = parents.get(LateralViewJoinOperator.SELECT_TAG).getStatistics();
       final Statistics udtfStats = parents.get(LateralViewJoinOperator.UDTF_TAG).getStatistics();
 
-      final double factor = (double) udtfStats.getNumRows() / (double) selectStats.getNumRows();
+      final long udtfNumRows = Math.max(udtfStats.getNumRows(), 1);
+      final double factor = (double) udtfNumRows / (double) Math.max(selectStats.getNumRows(), 1);
```
Comment (okumin) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:2965]: `factor` will be greater than 0.0 and must not 0 or infinity.

**Comment by okumin** [comment]
@kgyrtkirk I have updated some points so that # of rows will never be 0.
Could you please have a look when you have a chance?

**Review by kgyrtkirk — APPROVED** [review]
